contracts/swap/swap.go

```diff
@@ -39,44 +39,55 @@ var (
 	ErrTransactionReverted = errors.New("Transaction reverted")
 )
 
-// Validator struct -> put validator in implementation of Swap. Make the validator a package level function and implement this in Swap
-
 // Backend wraps all methods required for contract deployment.
 type Backend interface {
 	bind.ContractBackend
 	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
 	//TODO: needed? BalanceAt(ctx context.Context, address common.Address, blockNum *big.Int) (*big.Int, error)
 }
 
-// SimpleSwap interface defines the simple swap's exposed methods
-type SimpleSwap interface {
-	Deploy(auth *bind.TransactOpts, backend bind.ContractBackend, owner common.Address, harddepositTimeout *big.Int) (common.Address, *types.Transaction, error)
-	SubmitChequeBeneficiary(opts *bind.TransactOpts, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Transaction, error)
-	CashChequeBeneficiary(auth *bind.TransactOpts, backend Backend, beneficiary common.Address, requestPayout *big.Int) (*types.Transaction, error)
-	ValidateCode() bool
+// Deploy creates a new instance of the contract and immediately deploys it
+func Deploy(auth *bind.TransactOpts, backend bind.ContractBackend, owner common.Address, harddepositTimeout time.Duration) (common.Address, Contract, *types.Transaction, error) {
+	addr, tx, s, err := contract.DeploySimpleSwap(auth, backend, owner, big.NewInt(int64(harddepositTimeout)))
+	c := simpleContract{instance: s}
+	return addr, c, tx, err
+}
+
+// InstanceAt creates a new instance of a contract at a specific address.
+// It assumes that there is an existing contract instance at the given address, or an error is returned
+// This function is needed to communicate with remote Swap contracts (e.g. sending a cheque)
+func InstanceAt(address common.Address, backend bind.ContractBackend) (Contract, error) {
+	simple, err := contract.NewSimpleSwap(address, backend)
+	c := simpleContract{instance: simple}
+	return c, err
+}
+
+// Contract interface defines the simple swap's exposed methods
+type Contract interface {
+	SubmitChequeBeneficiary(opts *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Receipt, error)
+	CashChequeBeneficiary(auth *bind.TransactOpts, backend Backend, beneficiary common.Address, requestPayout *big.Int) (*types.Receipt, error)
 	ContractParams() *Params
-	InstanceAt(address common.Address, backend bind.ContractBackend)
+	Issuer(opts *bind.CallOpts) (common.Address, error)
+	Cheques(opts *bind.CallOpts, addr common.Address) (*ChequeResult, error)
```

Comment:
@ralph-pichler why does the go binding return an inline `struct{...]` for the `Cheques` function instead of some defined type?

Reply:
That's probably because the getter on the solidity side actually returns the fields of the struct as tuple and that type has no specific name.

Reply:
So we can keep this as it is now right?

Reply:
Yes